### PR TITLE
Gpenverne fix/allow empty content on post and put requests

### PR DESCRIPTION
**Pull request by dunglas**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #781, #782 |
| License | MIT |
| Doc PR | n/a |

Follows #782. All credit go to @gpenverne.

### CODE REVIEW

File: features/crud.feature [hunk 1]
```diff
@@ -358,6 +358,9 @@ Feature: Create-Retrieve-Update-Delete
       "alias": null
     }
     """
+  Scenario: PUT request on custom operations without content
```
Comment (teohhanhui) [inline features/crud.feature]: This scenario is incorrect. This is not a custom operation, and I don't suppose it should return 200 when the semantics is undefined (for the default PUT operation).

Comment (dunglas) [inline features/crud.feature]: Indeed it is incorrect. But why a 400 if we allow empty requests for `PUT`, it should be allowed for all `PUT` operations isn't it?

**Comment by teohhanhui** [comment]
What does PUT with an empty body mean? The semantics are undefined. The
purpose of this PR is to skip deserialization of empty body for custom
operations, but it should return 400 for the default POST / PUT operations.

**Comment by dunglas** [comment]
Ok got what you mean. Not sure that we should add this in core. Maybe just a doc to explain how to do that would be enough? It's weird to have a different behavior for built-in and user-land operations.

**Comment by teohhanhui** [comment]
> It's weird to have a different behavior for built-in and user-land operations.

Ideally it should return 400 for both built-in and custom operations, unless the user handles it in the custom operation. But yeah, I'm not sure how that can be implemented. So probably we should merge this as is.

**Comment by kbsali** [comment]
@teohhanhui why should it return a 400? Sending empty content is *not* an error!

**Comment by teohhanhui** [comment]
@kbsali:

> why should it return a 400? Sending empty content is not an error!

It really depends on the content type. For example, an empty string mean is NOT valid JSON.

**Comment by jdeniau** [comment]
Hi @soyuka or @dunglas ,

What is the status of this PR ? empty content has been allowed on PUT request via #1274 .

@teohhanhui : You are right that empty string is an invalid JSON but POST method do send empty string as body if you don't specify one.

For me this should be iso between all "unsafe" methods (except DELETE), if you send an error you should send errors on PUT, POST, PATCH, etc., but if you don't send an error, you should not send error at all.

**Comment by soyuka** [comment]
@jdeniau I think it'll be resolved once every point I stated in #1282 are. Maybe we should close this in favor of #1282.

**Pull request closed without merging**